devito/types/dense.py

```diff
@@ -1454,3 +1453,160 @@ def parent(self):
         return self._parent
 
     _pickle_kwargs = Function._pickle_kwargs + ['parent']
+
+
+class TempFunction(DiscreteFunction):
+
+    """
+    Tensor symbol used to store an intermediate sub-expression extracted from
+    one or more symbolic equations.
+
+    Users should not instantiate this class directly. TempFunctions may be created
+    by Devito to store intermediate sub-expressions ("temporary values") when the
+    user supplies the `cire-ftemps` option to an Operator.
+
+    Unlike other DiscreteFunction types, TempFunctions do not carry data directly.
+    However, they can generate Functions to override the TempFunction at Operator
+    application time (see the Examples section below).
+
+    TempFunctions are useful if the user wants to retain control over the allocation
+    and deletion of temporary storage (by default, instead, Devito uses Arrays, which
+    are allocated and deallocated upon entering and exiting C-land, respectively).
+
+    Examples
+    --------
+    The `make` method makes the TempFunction create a new Function. For more info,
+    refer to TempFunction.make.__doc__.
+
+      .. code-block:: python
+
+        op = Operator(...)
+        cfuncs = [i for i in op.input if i.is_TempFunction]
+        kwargs = {i.name: i.make(grid.shape) for i in cfuncs}
+        op.apply(..., **kwargs)
+    """
+
+    is_TempFunction = True
+
+    def __init_finalize__(self, *args, **kwargs):
+        super().__init_finalize__(*args, **kwargs)
+
+        self._pointer_dim = kwargs.get('pointer_dim')
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        if pointer_dim not in dimensions:
+            # This is a bit hacky but it does work around duplicate dimensions when
+            # it gets to pickling
+            dimensions = as_tuple(pointer_dim) + dimensions
+
+        # Sanity check
+        assert not any(d.is_NonlinearDerived for d in dimensions)
+
+        return dimensions, dimensions
+
+    def __halo_setup__(self, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        halo = as_tuple(kwargs.get('halo'))
+        if halo is None:
+            halo = tuple((0, 0) for _ in dimensions)
+        if pointer_dim is not None:
```

Comment:
what if halo is not None and a value was passed for `pointer_dim`'s halo

Reply:
that'd subsume a compiler error, but fair enough, I'll patch the line below now